### PR TITLE
Feature:  MergeManyChangeSets for Cache ChangeSets with List ChangeSets

### PR DESCRIPTION
**Pull request by dwcullop**
## Description

Implements the `MergeManyChangeSets` operator for Cache ChangeSets source and List ChangeSet children.  This PR also adds an `IEqualityComparer` to the List-to-List version that was omitted before.

## Other Changes

- Rename `MergeManyCacheChangeSetsSourceCompareFixture` to `MergeManyChangeSetsCacheSourceCompareFixture` for consistency
- Rename `Completed` to `IsCompleted` property on Cache Aggregator for consistency
- Created a Stress Test helper extension methods and test helper code

### CODE REVIEW

**Review by RolandPheasant — CHANGES_REQUESTED** [review]

**Comment by RolandPheasant** [comment]
Code wise, it looks great and only one minor request which is to seal classes when possible.

Aside from that I ran the stress tests and managed to produce some intermittent errors.

image

and on a subsequent run 

image

I think the stress tests are a brilliant idea, but I wonder whether that test performance wizard @JakenVeina has any suggested how those tests can be made to run fast,

**Comment by JakenVeina** [comment]
> I think the stress tests are a brilliant idea, but I wonder whether that test performance wizard @JakenVeina has any suggestions how those tests can be made to run fast,

Would be happy to, if the branch would build. ;)

File: src/DynamicData/List/Internal/MergeManyListChangeSets.cs [hunk 1]
```diff
@@ -7,7 +7,10 @@
 
 namespace DynamicData.List.Internal;
 
-internal class MergeManyListChangeSets<TObject, TDestination>(IObservable<IChangeSet<TObject>> source, Func<TObject, IObservable<IChangeSet<TDestination>>> selector)
+/// <summary>
+/// Operator that is similiar to MergeMany but intelligently handles List ChangeSets.
+/// </summary>
+internal class MergeManyListChangeSets<TObject, TDestination>(IObservable<IChangeSet<TObject>> source, Func<TObject, IObservable<IChangeSet<TDestination>>> selector, IEqualityComparer<TDestination>? equalityComparer = null)
```
[the remark this answers is not in the document]
Comment (ChrisPulman) [reply src/DynamicData/List/Internal/MergeManyListChangeSets.cs:13]: This has been resolved

**Comment by ChrisPulman** [comment]
The problem now is that the Stress tests take an infinite period of time to run.

File: src/DynamicData.Tests/Cache/ToObservableChangeSetFixture.cs [hunk 2]
```diff
@@ -403,7 +246,9 @@ public void SizeLimitIsExceeded_OldestItemsAreRemoved()
         scheduler.AdvanceBy(1);
 
         results.Error.Should().BeNull();
-        results.Messages.Count.Should().Be(6, "6 item sets were emitted by the source");
+
+        // TODO: This was set to 9 but fails, this was changed in a recent commit form 6 to 9, but I'm not sure why.
+        results.Messages.Count.Should().Be(6, "6 item sets were emitted by the source, 3 of which triggered followup evictions");
```
Comment (ChrisPulman) [inline src/DynamicData.Tests/Cache/ToObservableChangeSetFixture.cs:251]: Please check why this was changed from 6 to 9, I reverted the value to 6 to make the test pass but need to be sure what the correct value should be

**Comment by ChrisPulman** [comment]
> > I think the stress tests are a brilliant idea, but I wonder whether that test performance wizard @JakenVeina has any suggestions how those tests can be made to run fast,
> 
> Would be happy to, if the branch would build. ;)

@JakenVeina The branch should build now, if you have time

File: src/DynamicData.Tests/Cache/ToObservableChangeSetFixture.cs [hunk 3]
```diff
@@ -403,11 +246,12 @@ public void SizeLimitIsExceeded_OldestItemsAreRemoved()
         scheduler.AdvanceBy(1);
 
         results.Error.Should().BeNull();
-        results.Messages.Count.Should().Be(6, "6 item sets were emitted by the source");
+        results.Messages.Count.Should().Be(9, "6 item sets were emitted by the source, 3 of which triggered followup evictions");
         results.Data.Items.Should().BeEquivalentTo(new[] { item5, item6, item9, item10, item11 }, "the size limit of the collection was 5");
     }
 
-    [Fact]
+    [Fact(Skip = "Outstanding bug, notifications are not synchronized, initial item emits after error")]
```
[the remark this answers is not in the document]
Comment (dwcullop) [reply src/DynamicData.Tests/Cache/ToObservableChangeSetFixture.cs]: Not sure the status of this.  @ChrisPulman added that change to make the unit tests pass, but I'm not sure why they were failing.

**Comment by dwcullop** [comment]
> Code wise, it looks great and only one minor request which is to seal classes when possible.

Agree about sealing the classes.  Will do.

> Aside from that I ran the stress tests and managed to produce some intermittent errors.

Been fighting these for a few days now, but I hope I've got them under control now.
 
> I think the stress tests are a brilliant idea, but I wonder whether that test performance wizard @JakenVeina has any suggestions how those tests can be made to run fast,

I'm open to any suggestions.

File: src/DynamicData.Tests/Cache/ToObservableChangeSetFixture.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrisPulman) [reply src/DynamicData.Tests/Cache/ToObservableChangeSetFixture.cs]: SourceErrorsImmediately_SubscriptionReceivesError has been re-enabled I ran it on repeat for an extensive period with no failures, it was Skipped on the 9th Dec (First drop of MergeManyChangeSets for List to Cache) after some major reworks to the tests in this file. I have a feeling the test may have been skipped to then come back and fix the issue causing the failure, which looks like Darrin has successfully done.

File: src/DynamicData.Tests/Utilities/AddRemoveStressTester.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Linq;
+using System.Reactive;
+using System.Reactive.Concurrency;
+using System.Reactive.Disposables;
+using System.Threading;
+
+namespace DynamicData.Tests.Utilities;
+
+internal static class AddRemoveStressTester
```
[the remark this answers is not in the document]
Comment (ChrisPulman) [reply src/DynamicData.Tests/Utilities/AddRemoveStressTester.cs]: This has been resolved

**Review by RolandPheasant — APPROVED** [review]
I have not followed the entire conversation on Slack, owning to being away for a couple days. Instead for now I'll approve and leave to the discretion of @dwcullop whether it is ready to be merged.

**Pull request merged**